src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs

```diff
@@ -47,19 +47,50 @@ public static TypeSyntax AsReturnType(this IMarshallingGenerator generator, Type
             return AttributeList(SingletonSeparatedList(marshalAsAttribute));
         }
 
+        private const string ParameterIdentifierSuffix = "param";
+
         /// <summary>
         /// Gets a parameter for the unmanaged signature that represents the provided <paramref name="info"/>.
         /// </summary>
         /// <param name="generator">The marshalling generator for this <paramref name="info"/></param>
         /// <param name="info">Object to marshal</param>
-        public static ParameterSyntax AsParameter(this IMarshallingGenerator generator, TypePositionInfo info)
+        public static ParameterSyntax AsParameter(this IMarshallingGenerator generator, TypePositionInfo info, StubCodeContext context)
         {
             SignatureBehavior behavior = generator.GetNativeSignatureBehavior(info);
             if (behavior == SignatureBehavior.ManagedTypeAndAttributes)
             {
                 return GenerateForwardingParameter(info);
             }
-            return Parameter(Identifier(info.InstanceIdentifier))
+            string identifierName;
+            if (context.Direction == MarshalDirection.ManagedToUnmanaged)
+            {
+                // This name doesn't get introduced into the stub's scope, so we can make it pretty
+                // and reuse the native identifier
+                identifierName = context.GetIdentifiers(info).native;
+            }
+            else if (context.Direction == MarshalDirection.UnmanagedToManaged)
+            {
+                // This name is introduced into the stub's scope.
+                // When we are passing the managed identifier as-is, we can just use that name everywhere.
+                // When we're passing the native identifier as-is or casting the value to the native type in managed->unmanaged cases,
+                // we can use the native identifier.
+                // When we're passing the address of the native identifier, we need to introduce a new name to hold this value
+                // before we assign it to the managed value.
+                (string managed, string native) = context.GetIdentifiers(info);
+                string param = context.GetAdditionalIdentifier(info, ParameterIdentifierSuffix);
+                identifierName = generator.GetValueBoundaryBehavior(info, context) switch
+                {
+                    ValueBoundaryBehavior.ManagedIdentifier => info.IsByRef ? param : managed,
+                    ValueBoundaryBehavior.NativeIdentifier or ValueBoundaryBehavior.CastNativeIdentifier => native,
+                    ValueBoundaryBehavior.AddressOfNativeIdentifier => param,
+                    _ => throw new UnreachableException()
+                };
+            }
+            else
+            {
+                throw new ArgumentException("Context direction must be ManagedToUnmanaged or UnmanagedToManaged");
```

Comment:
Is this user visible? I can't recall if we are localizing these errors.

Reply:
This is not user-visible. We generally haven't been localizing these errors.